File: tests/algorithms/transpiler/FS/conversions/roman_numerals.fs
// Generated 2025-08-07 10:31 +0700

exception Break
exception Continue

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let roman_values: int array = [|1000; 900; 500; 400; 100; 90; 50; 40; 10; 9; 5; 4; 1|]
let roman_symbols: string array = [|"M"; "CM"; "D"; "CD"; "C"; "XC"; "L"; "XL"; "X"; "IX"; "V"; "IV"; "I"|]
let rec char_value (c: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable c = c
    try
        if c = "I" then
            __ret <- 1
            raise Return
        if c = "V" then
            __ret <- 5
            raise Return
        if c = "X" then
            __ret <- 10
            raise Return
        if c = "L" then
            __ret <- 50
            raise Return
        if c = "C" then
            __ret <- 100
            raise Return
        if c = "D" then
            __ret <- 500
            raise Return
        if c = "M" then
            __ret <- 1000
            raise Return
        __ret <- 0
        raise Return
        __ret
    with
        | Return -> __ret
let rec roman_to_int (roman: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable roman = roman
    try
        let mutable total: int = 0
        let mutable i: int = 0
        while i < (String.length (roman)) do
            if ((i + 1) < (String.length (roman))) && ((char_value (string (roman.[i]))) < (char_value (string (roman.[i + 1])))) then
                total <- (total + (char_value (string (roman.[i + 1])))) - (char_value (string (roman.[i])))
                i <- i + 2
            else
                total <- total + (char_value (string (roman.[i])))
                i <- i + 1
        __ret <- total
        raise Return
        __ret
    with
        | Return -> __ret
let rec int_to_roman (number: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable number = number
    try
        let mutable num: int = number
        let mutable res: string = ""
        let mutable i: int = 0
        try
            while i < (Seq.length (roman_values)) do
                try
                    let value: int = _idx roman_values (i)
                    let symbol: string = _idx roman_symbols (i)
                    let factor: int = num / value
                    num <- ((num % value + value) % value)
                    let mutable j: int = 0
                    while j < factor do
                        res <- res + symbol
                        j <- j + 1
                    if num = 0 then
                        raise Break
                    i <- i + 1
                with
                | Continue -> ()
                | Break -> raise Break
        with
        | Break -> ()
        | Continue -> ()
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/algorithms/x/PHP/conversions/roman_numerals.php
<?php
ini_set('memory_limit', '-1');
function _intdiv($a, $b) {
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv($a, $b);
}
$roman_values = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1];
$roman_symbols = ['M', 'CM', 'D', 'CD', 'C', 'XC', 'L', 'XL', 'X', 'IX', 'V', 'IV', 'I'];
function char_value($c) {
  global $roman_values, $roman_symbols;
  if ($c == 'I') {
  return 1;
}
  if ($c == 'V') {
  return 5;
}
  if ($c == 'X') {
  return 10;
}
  if ($c == 'L') {
  return 50;
}
  if ($c == 'C') {
  return 100;
}
  if ($c == 'D') {
  return 500;
}
  if ($c == 'M') {
  return 1000;
}
  return 0;
}
function roman_to_int($roman) {
  global $roman_values, $roman_symbols;
  $total = 0;
  $i = 0;
  while ($i < strlen($roman)) {
  if ($i + 1 < strlen($roman) && char_value(substr($roman, $i, $i + 1 - $i)) < char_value(substr($roman, $i + 1, $i + 1 + 1 - ($i + 1)))) {
  $total = $total + char_value(substr($roman, $i + 1, $i + 1 + 1 - ($i + 1))) - char_value(substr($roman, $i, $i + 1 - $i));
  $i = $i + 2;
} else {
  $total = $total + char_value(substr($roman, $i, $i + 1 - $i));
  $i = $i + 1;
}
};
  return $total;
}
function int_to_roman($number) {
  global $roman_values, $roman_symbols;
  $num = $number;
  $res = '';
  $i = 0;
  while ($i < count($roman_values)) {
  $value = $roman_values[$i];
  $symbol = $roman_symbols[$i];
  $factor = _intdiv($num, $value);
  $num = $num % $value;
  $j = 0;
  while ($j < $factor) {
  $res = $res . $symbol;
  $j = $j + 1;
};
  if ($num == 0) {
  break;
}
  $i = $i + 1;
};
  return $res;
}
